doc/man1/roarfilt.1
.\" roard.1:

.TH "roarfilt" "1" "July 2008" "RoarAudio" "System Manager's Manual: roarfilt"

.SH NAME

roarfilt \- RoarAudio sound server audio filter client

.SH SYNOPSIS

roarfilt [OPTIONS]...

.SH "DESCRIPTION"
This filters the audio data before they are send to the soundcard.

.SH "GENERAL OPTIONS"
\fBNOTE: This list is incomplete!\fR
In addition to this list this program provides compatible
options for esdfilt. You can just replace the "esd"
in the name with "roar" or use symlinks. This options are for compatibly
only and should not used in new programs.

.TP
\fB--help\fR
Print some help text.

.SH "AUDIO OPTIONS"
.TP
\fB--rate RATE\fR, \fB-R RATE\fR
Set server sample rate.

.TP
\fB--bits BITS\fR, \fB-B BITS\fR
Set server bits.

.TP
\fB--chans CHANNELS\fR, \fB-C CHANNELS\fR
Set server channels.

.SH "NETWORK OPTIONS"
.TP
\fB--server SERVER\fR
The server to connect to.

.SH "OLD ROARFILT OPTIONS"
.TP
\fB--half\fR
Half the volume.

.TP
\fB--double\fR
Double the volume.

.TP
\fB--amp n\fR, \fB--mul n\fR
Amplify volume by n.

.TP
\fB--div n\fR
Divide volume by n.

.TP
\fB--log n\fR
Does a simple 'gamma correction'.
This is done by:
 sample = 2^(bits-1) * log(1 + (n - 1)*sample/2^(bits-1)) / log(n).

.TP
\fB--lowpass freq\fR
lowpass filter

.SH "FILTER OPTIONS"

.TP
\fB--filter name\fR
This adds a new filter of name \fIname\fR.

.TP
\fB--ffreq freq\fR
This sets a frequency for the filter.

.TP
\fB--fmul mult\fR
This sets a multiplier for the filter.

.TP
\fB--fdiv div\fR
This sets a divider for the filter.

.TP
\fB--fn N\fR
This sets the filter's N value.

.TP
\fB--flimit limit\fR
This sets the filter's limit value.

.TP
\fB--fmode mode\fR
set filter mode parameter
This sets the filter mode. Symbolic values are currently not supported.

.TP
\fB--fq Q\fR
This sets the filter quality.


.SH "ENVIRONMENT VARIABLES"
Bellow only the basic environment varibales are listend.
For a complete list see \fBlibroar\fR(7).

.TP
\fBHOME\fR
The users home directory.

.TP
\fBROAR_SERVER\fR
The address of the listening server. This may be in form of host:port for TCP/IP connections
and /path/to/sock for UNIX Domain Sockets or any other supported format. See \fBroartips\fR(7)
for a full list.
This is the same as the \fB--server\fR option.

.SH "BUGS"
There are a lot of bugs...

.SH "SEE ALSO"
\fBroarctl\fR(1),
\fBroarvorbis\fR(1),
\fBroarfish\fR(1),
\fBroartestclients\fR(1),
\fBroartips\fR(7),
\fBlibroar\fR(7),
\fBRoarAudio\fR(7).

.SH "HISTORY"

For history information see \fBRoarAudio\fR(7).

.\" ll


ChangeLog
v. 0.2rc1 - ?
	* Started to port to OpenSolaris

v. 0.2beta3 - Fri May 01 2009 15:15
	* added esdfilt options to roarfilt, esdplay options to roarcatplay
	* added short options to roarcat/mon/filt as stated in the manpage
	* updated manpages
	* also build libroar{,dsp}.a
	* Changed VIO CTLs in a binary inkompatible way
	* added support for autoconf flag to OSS driver
	* added driver for OpenBSD sndio interface
	* started libroarsndio
	* added code for plugins: audacirus, mplayer, gst10
	* fixed some building problems
	* added roarify and s start for it's manpage
	* added stream flags pause and hwmixer

v. 0.2beta2 - Fri Mar 27 2009 21:20
	* added (dummy) flags autoconf and cleanmeta
	* added streams_ctl() to server
	* added support for primary flag as -oO, support for blocks parameter
	* added (GnuPG) pinentry support.
	* added support for Ogg Vorbis encoder to set meta data :)
	* fixed some (minor) bugs, including segfault in vio
	* added a lot things to the vio interface
	* added gzip and simple OpenPGP support to vio interface
	* added a simple pointer stack.
	* added (OpenSSH) askpass support.
	* added vio pipe support, currectly pipe() and socketpair() based onse.
	* test for OpenSSL
	* added support to VIO open OpenSSL BIOs
	* added support to read a string and convert it into a VIO stack
	* added simple VIO stack as VIO pass object
	* started cleaner VIO Socket implementation
	* Reduced lib size with --minimal trasticly (>33%!)
	* Done a lot patches to makeing RoarAudio more portable (also to win32)
	* moved convert.[ch] and poly.[ch] from libroar to libroardsp
	* updated AUTHROS, README and check HACKING
	* added tools/send-roaraudio: tool to send RoarAudio von CVS to a diffrent
	                              host in order to test it.
	* added dist/archlinux and archlinux rc script.

v. 0.2beta1 - Mon Feb 23 2009 25:08
	* added options --default-codec and --without-cf-cmd to configure script
	* added some funcs to list of frozen API calls
	* added sleep command to roarctl
	* updated roard manpage
	* fixed a possible segfault within auto codec detection and
	  added support for native flac
	* fixed some compiler warnings
	* made roarcat and roarmon 100% esdcat and esdmon compatible
	* set fragment size for OSS driver
	* fixed some minor bugs
	* changed buffer interface a bit to support non buffer alloced buffers

v. 0.2beta0 - Wed Feb 04 2009
	* use new driver interface as default (finally!)
	* fixed a bug in roard to not accept new streams in termination state
	* added support to filters to reset
	* added basic support to report latency
	* added vio support for ctl()s
	* allow overruns in sync output streams
	* solved double close bug
	* added support for non-execed streams with client behinde a NAT gateway
	* use socketpair() for AF_UNIX streams to create a new non-execed stream
	* got AF_UNIX support back working on OpenBSD
	* added needed paths for OpenBSD
	* only use use fsync() if there is no fdatasync()

v. 0.1beta5 - Wed Jan 21 2009
	* done some patchs to help porting libroar to win32
	* made libm optional
	* added option to configure to build without any optional compontents
	* added option to configure to give a helper programm to exec tests (e.g. wine)
	* updated help a bit
	* fixed roarbidir
	* added support for sync vio
	* updated roarfctest
	* added filter reset command

v. 0.1beta4 - Tue Dec 23 2008
	* added roar_buffer_shift_out()
	* changed roarshout to be 100% compatible with oggfwd options
	* added full support for mu-Law
	* added VIO support for driver API
	* changed esd output driver to use new vio driver API
	* wrote OSS driver
	* added support for --standby and --auto-standby
	* write support for stream based drivers, finally!
	* added -o/-O-/-oO/-oN to roard: stream based output drivers
	* patched null and raw driver to use VIO interface
	* ported libao plugin to new VIO interface
	* added -oP support to mark outputs primary
	* added ROAR_INSTINT to abstract from __LP64__
	* added new trivial filters: Add, Quantifier, Clip
	* added codecfilter for libfishsound, use it as default for ogg_flac and ogg_speex
	* added Ogg FLAC and Ogg Speex Magic
	* added tool roarmonhttp: http streaming tool, CGI or inetd
	* added writing support for RIFF/WAVE
	* flag sources as sources
	* patched some things for cygwin
	* fixed double-reconnect bug within XMMS plugin
	* include libao plugin in normal configure/build/install
	* added passthru for meta streams
	* fixed a lot small bugs
	* include basic rc/init script for debian like systems

v. 0.1beta3 - Thu Nov 27 2008
	* fixed some includes, patch by janus

v. 0.1beta2 - Tue Nov 25 2008
	* check for __LP64__
	* added roaraudio/error.h
	* added manpages for roar_codec2str() and roar_str2codec()
	* got some YIFF clients working, including yplay!
	* added roarcleints/roarradio*: A RoarAudio client which inizialiszes a new source
	                                and has buildin HTTP support.
	* added CDRum support
	* added writeing support for A-Law
	* added proxy support for SOCKS4 and 4d, SSH and HTTP CONNECT
	* improved 32 bit support a lot
	* checking for IPv6 support, not just assuming it to exist (win32 fix)
	* updated docs a bit
	* added install target

v. 0.1beta1 - Sun Sep 07 2008
	* fixed a lot typos
	* Added readonly support for A-Law
	* now also build doc/
	* fixed a bug with NetBSD's which
	* look for pkg-config
	* look for other tools, added --runtime-detect to configure
	* Fixed some manpages
	* added support for passive playback (passfh)
	* added support for ROAR_ATTACH_SIMPLE
	* cleaned codec ID #defines a bit up
	* added CMD PASSFH
	* get roard build again on MAC OS X
	* added tool roarcatpassfh
	* added --simple, --passive and --background to roarcatplay
	* got cf vorbis work on vio
	* got standard vio work better on DECnet

v. 0.1 - Sun Aug 31 2008
	* done basic stuff
	* started 'libroaresd' to support ESD clients
	* started 'libroararts' to support aRts clients

roarclients/roarfilt.c
//roarfilt.c:

/*
 *      Copyright (C) Philipp 'ph3-der-loewe' Schafft - 2008
 *
 *  This file is part of roarclients a part of RoarAudio,
 *  a cross-platform sound system for both, home and professional use.
 *  See README for details.
 *
 *  This file is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License version 3
 *  as published by the Free Software Foundation.
 *
 *  RoarAudio is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this software; see the file COPYING.  If not, write to
 *  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#include <roaraudio.h>
#include <libroardsp/libroardsp.h>

#ifdef ROAR_HAVE_LIBM
#include <math.h>
#endif

#define BUFSIZE 1024
#ifdef ROAR_HAVE_LIBM
struct {
 uint16_t a, b;
 int16_t  old[ROAR_MAX_CHANNELS];
} g_lowpass;
#endif

void usage (void) {
 printf("roarfilt [OPTIONS]...\n");

 printf("\nOptions:\n\n");

 printf("  --server SERVER    - Set server hostname\n"
        "  --rate   RATE      - Set sample rate\n"
        "  --bits   BITS      - Set bits per sample\n"
        "  --chans  CHANNELS  - Set number of channels\n"
        "  --help             - Show this help\n"
        "\n"
        "  --half             - half the volume\n"
        "  --double           - double the volume\n"
        "  --amp VAL          - Set amplification\n"
        "  --mul VAL          - Set mul\n"
        "  --div VAL          - Set div\n"
#ifdef ROAR_HAVE_LIBM
        "  --lowpass freq     - lowpass filter\n"
#endif
        "  --filter  name     - add filter name\n"
        "  --ffreq   freq     - set filter freq\n"
        "  --fmul    mult     - set filter multiplier\n"
        "  --fdiv    div      - set filter divider\n"
        "  --fn      N        - set filter N parameter\n"
        "  --flimit  limit    - set filter limit parameter\n"
        "  --fmode   mode     - set filter mode parameter\n"
        "  --fq      Q        - set filter quality\n"
       );

}

void vol2 (void * data, int mul, int div, int len) {
 int16_t * samples = (int16_t *) data;
 int i;

 len /= 2;

 for (i = 0; i < len; i++)
  samples[i] = ((int) samples[i] * mul) / div;
}

void vol1 (void * data, int mul, int div, int len) {
 int8_t * samples = (int8_t *) data;
 int i;

 for (i = 0; i < len; i++)
  samples[i] = ((int) samples[i] * mul) / div;
}

#ifdef ROAR_HAVE_LIBM
void logs2 (void * data, float scale, int len) {
 int16_t * samples = (int16_t *) data;
 int i;
 float div = logf(scale);
 float scalemul = scale - 1;
 int neg;

 len /= 2;

 //printf("logs2(data=%p, scale=%f, len=%i): scalemul=%f, div=%f\n", data, scale, len, scalemul, div);

 for (i = 0; i < len; i++) {
  if ( (neg = (samples[i] < 0)) ) 
   samples[i] = abs(samples[i]);


  samples[i] = (neg ? 32768.0 : 32767.0)*logf(1 + (scalemul*(float)samples[i]/(neg ? 32768.0 : 32767.0))) / div;

  if ( neg )
   samples[i] *= -1;
 }
}

void lowpass2 (void * data, int len, int channels) {
 int16_t * samples = (int16_t *) data;
 register int32_t s;
 int i, c;

 if ( channels > ROAR_MAX_CHANNELS )
  return;

 len /= 2 * channels;

//  *      output[N] = input[N] * A + output[N-1] * B

 for (i = 0; i < len; i++) {
  for (c = 0; c < channels; c++) {
   s = samples[i*channels + c] * g_lowpass.a + g_lowpass.old[c] * g_lowpass.b;

   s /= 65536;

   samples[i*channels + c] = s;
   g_lowpass.old[       c] = s;
  }
 }
}
#endif

int main (int argc, char * argv[]) {
 int    rate     = 44100;
 int    bits     = 16;
 int    channels = 2;
 int    codec    = ROAR_CODEC_DEFAULT;
 char * server   = NULL;
 char * k;
 int    fh;
 int    i;
 int    mul = 1, div = 1;
 int32_t tmp;
 float  logscale = 0;
 float  lp       = 0;
 char buf[BUFSIZE];
 struct roardsp_filterchain fc;
 struct roardsp_filter      filter_real[8];
 struct roardsp_filter    * filter = filter_real - 1;
 struct roar_stream         stream;

#ifdef ROAR_HAVE_LIBM
 memset(&g_lowpass, 0, sizeof(g_lowpass));
#endif

 roardsp_fchain_init(&fc);

 for (i = 1; i < argc; i++) {
  k = argv[i];

  if ( strcmp(k, "--server") == 0 || strcmp(k, "-s") == 0 ) {
   server = argv[++i];
  } else if ( strcmp(k, "--rate") == 0 || strcmp(k, "-R") == 0 || strcmp(k, "-r") == 0 ) {
   rate = atoi(argv[++i]);
  } else if ( strcmp(k, "--bits") == 0 || strcmp(k, "-B") == 0 ) {
   bits = atoi(argv[++i]);
  } else if ( strcmp(k, "--channels") == 0 || strcmp(k, "--chans") == 0 || strcmp(k, "-C") == 0 ) {
   channels = atoi(argv[++i]);
  } else if ( strcmp(k, "-b") == 0 ) {
   bits = 8;
  } else if ( strcmp(k, "-m") == 0 ) {
   channels = 1;
  } else if ( strcmp(k, "--half") == 0 || strcmp(k, "-half") == 0 ) {
   div *= 2;
  } else if ( strcmp(k, "--double") == 0 || strcmp(k, "-double") == 0 ) {
   mul *= 2;
  } else if ( strcmp(k, "--amp") == 0 ) {
   mul *= atoi(argv[++i]);
  } else if ( strcmp(k, "--mul") == 0 ) {
   mul  = atoi(argv[++i]);
  } else if ( strcmp(k, "--div") == 0 ) {
   div  = atoi(argv[++i]);
  } else if ( strcmp(k, "--log") == 0 ) {
   logscale = atof(argv[++i]);
#ifdef ROAR_HAVE_LIBM
  } else if ( strcmp(k, "--lowpass") == 0 ) {
   lp = exp(-2 * M_PI * atof(argv[++i]) / rate) * 65536;
   g_lowpass.b = lp;
   g_lowpass.a = 65536 - lp;
#endif
//   printf("lowpass: A=%i, B=%i\n", g_lowpass.a, g_lowpass.b);
  } else if ( strcmp(k, "--filter") == 0 ) {
   stream.info.channels = channels;
   stream.info.bits     = bits;
   stream.info.rate     = rate;
   filter++;
   roardsp_filter_init(filter, &stream, roardsp_filter_str2id(argv[++i]));
   roardsp_fchain_add(&fc, filter);
  } else if ( strcmp(k, "--ffreq") == 0 ) {
   lp = atof(argv[++i]);
   roardsp_filter_ctl(filter, ROARDSP_FCTL_FREQ, &lp);
  } else if ( strcmp(k, "--fmul") == 0 ) {
   tmp = atoi(argv[++i]);
   roardsp_filter_ctl(filter, ROARDSP_FCTL_MUL, &tmp);
  } else if ( strcmp(k, "--fdiv") == 0 ) {
   tmp = atoi(argv[++i]);
   roardsp_filter_ctl(filter, ROARDSP_FCTL_DIV, &tmp);
  } else if ( strcmp(k, "--fn") == 0 ) {
   tmp = atoi(argv[++i]);
   roardsp_filter_ctl(filter, ROARDSP_FCTL_N, &tmp);
  } else if ( strcmp(k, "--fq") == 0 ) {
   tmp = atoi(argv[++i]);
   roardsp_filter_ctl(filter, ROARDSP_FCTL_Q, &tmp);
  } else if ( strcmp(k, "--flimit") == 0 ) {
   tmp = atoi(argv[++i]);
   roardsp_filter_ctl(filter, ROARDSP_FCTL_LIMIT, &tmp);
  } else if ( strcmp(k, "--fmode") == 0 ) {
   tmp = atoi(argv[++i]);
   roardsp_filter_ctl(filter, ROARDSP_FCTL_MODE, &tmp);
  } else if ( strcmp(k, "--help") == 0 || strcmp(k, "-h") == 0 ) {
   usage();
   return 0;
  } else {
   fprintf(stderr, "Error: unknown argument: %s\n", k);
   usage();
   return 1;
  }
 }

 if ( (fh = roar_simple_filter(rate, channels, bits, codec, server, "roarifilt")) == -1 ) {
  fprintf(stderr, "Error: can not start playback\n");
  return 1;
 }

 if ( mul == div &&
#ifdef ROAR_HAVE_LIBM
      logscale == 0 && g_lowpass.a == 0 &&
#endif
      roardsp_fchain_num(&fc) == 0 ) {
  fprintf(stderr, "Error: filter is useless!\n");
  return 0;
 }

 if ( bits == 16 ) {
  while((i = read(fh, buf, BUFSIZE))) {
   if ( mul != div )
    vol2((void*)buf, mul, div, i);
#ifdef ROAR_HAVE_LIBM
   if ( logscale )
    logs2((void*)buf, logscale, i);
   if ( g_lowpass.a )
    lowpass2((void*)buf, i, channels);
#endif
   roardsp_fchain_calc(&fc, (void*)buf, (8*i)/bits);
   if (write(fh, buf, i) != i)
    break;
  }
 } else if ( bits == 8 ) {
  while((i = read(fh, buf, BUFSIZE))) {
   vol1((void*)buf, mul, div, i);
   if (write(fh, buf, i) != i)
    break;
  }
 } else {
  fprintf(stderr, "Error: %i bits per sample is not supported!\n", bits);
  return 1;
 }

 roar_simple_close(fh);

 roardsp_fchain_uninit(&fc);

 return 0;
}

//ll
